packages/bridge-ui-v2/src/libs/proof/Prover.ts

```diff
@@ -1,50 +1,103 @@
 import { getContract, type GetContractResult, type PublicClient } from '@wagmi/core';
-import { type Address, encodePacked, type Hex, keccak256 } from 'viem';
+import { type Address, encodePacked, type Hex, keccak256, toHex } from 'viem';
 
 import { crossChainSyncABI } from '$abi';
 import { routingContractsMap } from '$bridgeConfig';
 import { PendingBlockError } from '$libs/error';
 import { getLogger } from '$libs/util/logger';
 import { publicClient } from '$libs/wagmi';
 
-import type { ClientWithEthGetProofRequest, GenerateProofArgs } from './types';
+import type { Block, ClientWithEthGetProofRequest, GenerateProofArgs } from './types';
 
 const log = getLogger('proof:Prover');
 
 export class Prover {
-  protected async _getKey(contractAddress: Address, msgHash: Hex) {
+  protected async _getKeyToClaim(contractAddress: Address, msgHash: Hex) {
     return keccak256(encodePacked(['address', 'bytes32'], [contractAddress, msgHash]));
   }
 
+  protected async _getKeyToRecall(msgHash: Hex) {
+    return keccak256(encodePacked(['bytes', 'bytes32'], [toHex('MESSAGE_STATUS'), msgHash]));
+  }
+
   protected async _getLatestBlock(
     client: PublicClient,
     crossChainSyncContract: GetContractResult<typeof crossChainSyncABI>,
-  ) {
+  ): Promise<Block> {
     const latestBlockHash = await crossChainSyncContract.read.getCrossChainBlockHash([BigInt(0)]);
-    return client.getBlock({ blockHash: latestBlockHash });
+
+    return await client.request({
+      method: 'eth_getBlockByHash',
+      params: [
+        latestBlockHash,
+        true
+      ],
+    });
   }
 
-  async generateProof(args: GenerateProofArgs) {
-    const { msgHash, clientChainId, contractAddress, crossChainSyncChainId, proofForAccountAddress } = args;
 
-    const crossChainSyncAddress = routingContractsMap[crossChainSyncChainId][clientChainId].crossChainSyncAddress;
+
+  async generateRecallProof(args: GenerateProofArgs) {
+    const { msgHash, srcChainId, destChainId, proofForAccountAddress } = args;
+    const key = await this._getKeyToRecall(msgHash);
+    log('key to recall', key)
+
+    const destClient = publicClient({ chainId: destChainId });
+    const crossChainSyncAddress = routingContractsMap[srcChainId][destChainId].crossChainSyncAddress;
+
+    const crossChainSyncContract = getContract({
+      chainId: srcChainId,
+      address: crossChainSyncAddress,
+      abi: crossChainSyncABI,
+    });
+
+    const block = await this._getLatestBlock(destClient, crossChainSyncContract);
+    log('retrieved block', block);
+
+    const clientWithEthProofRequest = destClient as ClientWithEthGetProofRequest;
+
+    if (!block) throw new Error("could not fetch block")
+
+    if (!block || !block.hash) { throw new Error('block  is null') }
```

Comment:
Maybe we want to delete the line right above, which is already checking if the block is null or undefined?